Generic Cipher Generate, Update, and Transubstantiate

jgross11 opened the pull request:
aka The Great Polymorphic Disaster of 2021

jgross11 committed:
Cipher View Updates

Cipher views now use the cached cipher name / type list for generation

jgross11 committed:
Update Program.cs

cipher name, type, and derived JSON settings now created upon running program

jgross11 committed:
Update Utils.cs

Get / Set complex session object now properly handles derived classes

jgross11 committed:
Cipher class refactoring

Moved cipher classes to own namespace for startup type name generation convenience

jgross11 committed:
Create Options base + derived

Created front end form input types as classes to be stored in a cipher for fetching / updating form data

jgross11 committed:
Update CipherController.cs

Cipher controller getpage, encrypt, and decrypt now generalized (hopefully) for future cipher systems

jgross11 committed:
Update ProblemController.cs

Added problem creation submission for admins

jgross11 committed:
Update Cipher.cs

created mod helper method

jgross11 committed:
Create MASC.cs

started basic MASC implementation

andersonjackc commented on Project EFT/Ciphers/MASC.cs:
```diff
@@ -0,0 +1,79 @@
+                // otherwise, do shift
+                else
+                {
+                    ciphertext += substitutionAlphabet[index];
```
This does not handle the case where the original alphabet may be longer than the substitution alphabet, and vice versa for decrypt.  There is an index out of bounds exception thrown.

andersonjackc reviewed (changes requested):
Found small error with the MASC implementation, everything else looks great since we now have a generic cipher page.  Perhaps 1 small change would be the size of the textboxes for the alphabets, so that the entirety of the alphabet can be seen at one time.

jgross11 committed:
Merge branch 'main' into josh-pages

jgross11 committed:
Update Project EFT.csproj

added dependency for image manipulation package

jgross11 committed:
Added error messages

Implemented generation, formatting, etc

jgross11 committed:
Create RadioOptionsSet.cs

Created Option representing a set of options toggled via a radio button

jgross11 committed:
Create SimpleRadioStringOption.cs

Created option representing a simple string, to be used (probably) exclusively within a RadioOptionsSet

jgross11 committed:
Create ImageDisplayer.cs

Created Option representing an image whose source is a base64 data string

jgross11 committed:
Create ImageUploader.cs

Created Option representing a file uploader that only allows for image files to be uploaded

jgross11 committed:
Update CipherController.cs

Addition of RadioOptionsSet / ImageUploader called for a subtle change in generic form value setting - this is more flexible for future Options if needbe

jgross11 committed:
Radio Set Logic

Updated existing ciphers to allow (did not implement explicit decipherment methods) for variable decipherment method based on selected radio button on front end form

jgross11 committed:
Create VisualMASC.cs

Created cipher system that maps letters in the (standard lowercase) alphabet to a shade of red, and grammar / unknown characters to blue and vice versa

jgross11 committed:
Update ImageDisplayer.cs

cheap fix for short texts generating unseeable images, also ensures book text images don't fill the entire screen

jgross11 committed:
Update VisualMASC.cs

modified decryption for (what should be) much more efficient pixel color reading, but the gains are minimal :(

jgross11 committed:
mission critical bug fix

removed extra spaces

jgross11 committed:
Update VisualMASC.cs

- ensured uppercase letters are translated (to lowercase)
- fixed red out of range issue
- fixed null (0,0,0,0) color issue

andersonjackc commented:
Nice work!  Glad you got the mission critical bug fix done :)
The images is really cool!
